service/src/main/java/bio/terra/tanagra/service/artifact/CohortService.java

```diff
@@ -127,6 +144,87 @@ public Cohort updateCohort(
     return cohortDao.getCohort(cohortId);
   }
 
+  /** Clone an existing cohort's most recent revision, including reviews and annotations */
+  public Cohort cloneCohort(
+      String studyId,
+      String cohortId,
+      String userEmail,
+      @Nullable String displayName,
+      @Nullable String description) {
+    Cohort original = getCohort(studyId, cohortId);
+
+    // CohortRevision
+    Map<String, CohortRevision> clonedRevisionMap = new HashMap<>(); // to create reviews
+    List<CohortRevision> clonedRevisionList = new ArrayList<>(); // to maintain original order
+    original
+        .getRevisions()
+        .forEach(
+            revision -> {
+              CohortRevision clonedRevision =
+                  CohortRevision.builder()
+                      .version(revision.getVersion())
+                      .setIsMostRecent(revision.isMostRecent())
+                      .setIsEditable(revision.isEditable())
+                      .createdBy(userEmail)
+                      .lastModifiedBy(userEmail)
+                      .recordsCount(revision.getRecordsCount())
+                      // CriteriaGroupSections (shallow copy): they are written to DB & fetched.
+                      // Any ids are used in conjunction with cohort_revision_id as primary key
+                      .sections(revision.getSections())
+                      .build();
+              clonedRevisionMap.put(original.getId(), clonedRevision);
+              clonedRevisionList.add(clonedRevision);
+            });
+
+    // Cohort
+    Cohort.Builder cohortBuilder =
+        Cohort.builder()
+            .underlay(original.getUnderlay())
+            .displayName(
+                displayName != null ? displayName : "Copy of: " + original.getDisplayName())
+            .description(
+                description != null ? description : "Copy of: " + original.getDescription())
+            .createdBy(userEmail)
+            .lastModifiedBy(userEmail)
+            .revisions(clonedRevisionList);
+
+    // CohortReview
+    ResourceCollection authorizedReviewIds =
+        accessControlService.listAuthorizedResources(
+            SpringAuthentication.getCurrentUser(),
+            Permissions.forActions(REVIEW, READ),
+            ResourceId.forCohort(studyId, cohortId),
+            0,
+            MAX_REVIEW_SIZE);
+    List<Review> clonedReviewList = new ArrayList<>();
+    reviewService
+        .listReviews(authorizedReviewIds, 0, MAX_REVIEW_SIZE)
+        .forEach(
+            review ->
+                clonedReviewList.add(
+                    Review.builder()
+                        .displayName("Copy of: " + review.getDisplayName())
+                        .description("Copy of: " + review.getDescription())
```

Comment:
I don't think we need to update the names/descriptions of all child objects also.

Reply:
Only made this for those with Ids (deep copy) / those i thought might be displayed to the user -- in this case the user can differentiate the original from the clone

Reply:
At least in the core Tanagra UI, the child objects (reviews, annotations) are only shows in the context of a cohort. So being able to differentiate the original/cloned cohort is important, but I don't think it matters for the child objects.

Reply:
sg, was not sure if other objects were displayed anywhere.